runners/core-java/src/main/java/org/apache/beam/runners/core/metrics/SimpleExecutionState.java

```diff
@@ -53,8 +86,40 @@ public long getTotalMillis() {
     return totalMillis;
   }
 
+  @VisibleForTesting
+  public String getLullMessage(Thread trackedThread, Duration millis) {
+    // TODO(ajamato): Share getLullMessage code with DataflowExecutionState.
+    String userStepName =
+        this.labelsMetadata.getOrDefault(SimpleMonitoringInfoBuilder.PTRANSFORM_LABEL, null);
+    StringBuilder message = new StringBuilder();
+    message.append("Processing stuck");
+    if (userStepName != null) {
+      message.append(" in step ").append(userStepName);
+    }
+    message
+        .append(" for at least ")
+        .append(formatDuration(millis))
+        .append(" without outputting or completing in state ")
+        .append(getStateName());
+    message.append("\n");
+
+    StackTraceElement[] fullTrace = trackedThread.getStackTrace();
+    for (StackTraceElement e : fullTrace) {
+      message.append("  at ").append(e).append("\n");
+    }
+    return message.toString();
+  }
+
   @Override
   public void reportLull(Thread trackedThread, long millis) {
-    // TOOD(ajamato): Implement lullz detection to log stuck PTransforms.
+    if (!LOG.isWarnEnabled()) {
```

Comment:
This check should be unnecessary, right? The logging framework will discard the message for you if the log level is not enabled.